inference/enterprise/workflows/execution_engine/compiler/graph_constructor.py

```diff
@@ -0,0 +1,602 @@
+            public_message=f"Graph definition contains selector {end} that points to not defined element.",
+            context="workflow_compilation | execution_graph_construction",
+        )
+
+
+def verify_output_selector_points_to_valid_output(
+    output_selector: str,
+    step_outputs: List[OutputDefinition],
+) -> None:
+    selected_output_name = get_last_chunk_of_selector(selector=output_selector)
+    if selected_output_name == "*":
+        return None
+    defined_output_names = {output.name for output in step_outputs}
+    if selected_output_name not in defined_output_names:
+        raise InvalidReferenceTargetError(
+            public_message=f"Graph definition contains selector {output_selector} that points to output of step "
+            f"that is not defined in workflow block used to create step.",
+            context="workflow_compilation | execution_graph_construction",
+        )
+
+
+def verify_each_node_reach_at_least_one_output(
+    execution_graph: DiGraph,
+    manifest_class2block_class: Dict[Type[WorkflowBlockManifest], Type[WorkflowBlock]],
+) -> None:
+    all_nodes = set(execution_graph.nodes())
+    output_nodes = get_nodes_of_specific_kind(
+        execution_graph=execution_graph, kind=OUTPUT_NODE_KIND
+    )
+    nodes_without_outputs = get_nodes_that_do_not_produce_outputs(
+        execution_graph=execution_graph,
+        manifest_class2block_class=manifest_class2block_class,
+    )
+    nodes_that_must_be_reached = output_nodes.union(nodes_without_outputs)
+    nodes_reaching_output = (
+        get_nodes_that_are_reachable_from_pointed_ones_in_reversed_graph(
+            execution_graph=execution_graph,
+            pointed_nodes=nodes_that_must_be_reached,
+        )
+    )
+    nodes_not_reaching_output = all_nodes.difference(nodes_reaching_output)
+    if len(nodes_not_reaching_output) > 0:
+        raise DanglingExecutionBranchError(
+            public_message=f"Detected {len(nodes_not_reaching_output)} nodes not reaching any of output node:"
+            f"{nodes_not_reaching_output}.",
+            context="workflow_compilation | execution_graph_construction",
+        )
+
+
+def get_nodes_that_do_not_produce_outputs(
+    execution_graph: DiGraph,
+    manifest_class2block_class: Dict[Type[WorkflowBlockManifest], Type[WorkflowBlock]],
+) -> Set[str]:
+    # assumption is that nodes without outputs will produce some side effect and shall be
+    # treated as output nodes while checking if there is no dangling steps in graph
+    step_nodes = get_nodes_of_specific_kind(
+        execution_graph=execution_graph, kind=STEP_NODE_KIND
+    )
+    result = set()
+    for step_node in step_nodes:
+        step_manifest = execution_graph.nodes[step_node][NODE_DEFINITION_KEY]
+        step_manifest_type = type(step_manifest)
+        block_type = manifest_class2block_class[step_manifest_type]
+        if len(block_type.get_actual_outputs(step_manifest)) == 0:
```

Comment:
`if len(block_type.get_actual_outputs(step_manifest)) == 0:` -> `if not block_type.get_actual_outputs(step_manifest)`

Reply:
👍